app/src/main/java/com/example/androidtrainingproject/ui/home/HomeViewModel.kt

```diff
@@ -51,11 +57,52 @@ class HomeViewModel @Inject constructor(
         val newFilteredProducts = allProducts.filter { product: ProductResponse ->
             product.title.contains(newValue, ignoreCase = true)
         }
-        filteredProducts = newFilteredProducts
+        filteredProducts.clear()
+        filteredProducts.addAll(newFilteredProducts)
     }
 
     fun clearSearchText() {
         _searchText.value = ""
-        filteredProducts = allProducts
+        filteredProducts.clear()
+        filteredProducts.addAll(allProducts)
+    }
+
+    fun applyFilters(
+        checkedCategories: MutableList<String>,
+        rating: Int,
+        price: ClosedFloatingPointRange<Float>
+    ) {
+        filteredProducts.clear()
+        filteredProducts.addAll(allProducts.filter { product ->
+            checkedCategories.contains(product.category)
+                && product.rating.toInt() >= rating
+                && product.price >= price.start && product.price <= price.endInclusive
+        })
+
+        getActiveFiltersCount(checkedCategories, rating, price)
+    }
+
+    private fun getActiveFiltersCount(
+        checkedCategories: MutableList<String>,
+        rating: Int,
+        price: ClosedFloatingPointRange<Float>
+    ) {
+        var count = 0;
+        if (rating != 1) {
```

Comment:
Tip: If the `if` condition is small you can one-line it without curly braces, i.e: `if (rating != 1) count += 1`